ports/stm32/machine_i2s.c

```diff
@@ -0,0 +1,1109 @@
+//   32 byte address boundary.  Not all STM32 devices have a D-Cache.  Buffer alignment
+//   will still happen on these devices to keep this code simple.
+
+#define MAX_I2S_STM32 (2)
+
+// DMA ping-pong buffer size was empirically determined.  It is a tradeoff between:
+// 1. memory use (smaller buffer size desirable to reduce memory footprint)
+// 2. interrupt frequency (larger buffer size desirable to reduce interrupt frequency)
+// The sizeof 1/2 of the DMA buffer must be evenly divisible by the cache line size of 32 bytes.
+#define SIZEOF_DMA_BUFFER_IN_BYTES (256)
+#define SIZEOF_HALF_DMA_BUFFER_IN_BYTES (SIZEOF_DMA_BUFFER_IN_BYTES / 2)
+
+// For non-blocking mode, to avoid underflow/overflow, sample data is written/read to/from the ring buffer at a rate faster
+// than the DMA transfer rate
+#define NON_BLOCKING_RATE_MULTIPLIER (4)
+#define SIZEOF_NON_BLOCKING_COPY_IN_BYTES (SIZEOF_HALF_DMA_BUFFER_IN_BYTES * NON_BLOCKING_RATE_MULTIPLIER)
+
+#define NUM_I2S_USER_FORMATS (4)
+#define I2S_RX_FRAME_SIZE_IN_BYTES (8)
+
+typedef enum {
+    MONO,
+    STEREO
+} format_t;
+
+typedef enum {
+    BLOCKING,
+    NON_BLOCKING,
+    UASYNCIO
+} io_mode_t;
+
+typedef enum {
+    TOP_HALF,
+    BOTTOM_HALF
+} ping_pong_t;
+
+typedef struct _ring_buf_t {
+    uint8_t *buffer;
+    size_t head;
+    size_t tail;
+    size_t size;
+} ring_buf_t;
+
+typedef struct _non_blocking_descriptor_t {
+    mp_buffer_info_t appbuf;
+    uint32_t index;
+    bool copy_in_progress;
+} non_blocking_descriptor_t;
+
+typedef struct _machine_i2s_obj_t {
+    mp_obj_base_t base;
+    uint8_t i2s_id;
+    mp_hal_pin_obj_t sck;
+    mp_hal_pin_obj_t ws;
+    mp_hal_pin_obj_t sd;
+    uint16_t mode;
+    int8_t bits;
+    format_t format;
+    int32_t rate;
+    int32_t ibuf;
+    mp_obj_t callback_for_non_blocking;
+    uint8_t dma_buffer[SIZEOF_DMA_BUFFER_IN_BYTES + 0x1f]; // 0x1f related to D-Cache alignment
+    uint8_t *dma_buffer_dcache_aligned;
+    ring_buf_t ring_buffer;
+    uint8_t *ring_buffer_storage;
+    non_blocking_descriptor_t non_blocking_descriptor;
+    io_mode_t io_mode;
+    I2S_HandleTypeDef hi2s;
+    DMA_HandleTypeDef hdma_tx;
+    DMA_HandleTypeDef hdma_rx;
+    const dma_descr_t *dma_descr_tx;
+    const dma_descr_t *dma_descr_rx;
+} machine_i2s_obj_t;
+
+// The frame map is used with the readinto() method to transform the audio sample data coming
+// from DMA memory (32-bit stereo) to the format specified
+// in the I2S constructor.  e.g.  16-bit mono
+STATIC const int8_t i2s_frame_map[NUM_I2S_USER_FORMATS][I2S_RX_FRAME_SIZE_IN_BYTES] = {
+    { 0,  1, -1, -1, -1, -1, -1, -1 },  // Mono, 16-bits
+    { 2,  3,  0,  1, -1, -1, -1, -1 },  // Mono, 32-bits
+    { 0,  1,  4,  5, -1, -1, -1, -1 },  // Stereo, 16-bits
+    { 2,  3,  0,  1,  6,  7,  4,  5 },  // Stereo, 32-bits
+};
+
+//  This is used by the DMA callback functions to access the I2S object
+STATIC machine_i2s_obj_t *machine_i2s_obj[MAX_I2S_STM32];
```

Comment:
This will probably need to go in the root pointer section, eg like `pyb_uart_obj_all`.  Then access it via `MP_STATE_PORT(machine_i2s_obj)`.

Reply:
I held off on this change as I don't understand the issue or risk.  I had considered adding this array to the root pointer section, but the array is holding references to I2S objects, which are allocated in the recommended way, and should not be prematurely GC'd.  I can see a different situation for "stranded" heap objects that are allocated using m_new() and held inside the C code, and not connected to any structure that the GC can see  - in that case I understand the need to declare a root pointer, so the "stranded" object is not GC'd.

Reply:
You are right that if the users code holds a reference to the object then they will not be GC'd.  The issue comes about if the user creates an object, registers an IRQ callback, then no longer holds on to it.  Then it can get GC'd and the IRQ will run with invalid data.  This is pretty easy to do with Pin and Timer IRQ callbacks.  It's probably also possible to do with an I2S object, but if you think it's ok as-is then let's leave it, and possibly change it later if the need arises.

Reply:
I hadn't considered the possibility of a race condition.  I can see that possibility in the I2S implementation.  I'll definitely make the root pointer fix that you recommend, in both the STM32 and ESP32 ports.  There is nothing more heinous than a non-reproducible crash.

Reply:
I started work on the root pointer fix, but it seems there are some tricky aspects to adding root pointers to  `esp32/mpconfigport.h`.  

For example, in the root pointer section I need to add a line for I2S - shown below.  To compile this I need a way to declare the macro `I2S_NUM_MAX `.  Unfortunately adding the relevant header file `#include driver/i2s.h` causes many compile errors.  Substituting `I2S_NUM_MAX ` with the magic number `2` gets everything compiling, but that's not a very maintainable solution.

Rather than continue reverse engineering the code, I thought it best to ask for help ... 

```
#define MICROPY_PORT_ROOT_POINTERS \
    const char *readline_hist[8]; \
    mp_obj_t machine_pin_irq_handler[40]; \
    struct _machine_timer_obj_t *machine_timer_obj_head; \
    struct _machine_i2s_obj_t *machine_i2s_obj[I2S_NUM_MAX]; \
    MICROPY_PORT_ROOT_POINTER_BLUETOOTH_NIMBLE
```

Reply:
Ok, interesting... some options:
1. make `machine_i2s_obj` a `void*` and in `i2s_init0()` allocate (on the GC heap) an array of the right size and type and point the root pointer to it
2. `#define MP_I2S_MAX 2` in `mpconfigport.h` and static assert that it's less than or equal to `I2S_NUM_MAX` somewhere in `machine_i2s.c`

Reply:
Thanks.  Using a compile-time assert is a good fit for this problem.  I'll move forward with option 2.

Reply:
I made a commit adding `machine_i2s_obj[]` to the root pointer section.  Builds and testing on my local machine worked, but two builds on the server failed after commit.  I took quick look and one is easy to fix (nucleo build), but the other build (ESP32/IDF) shows a non-obvious failure.  Unfortunately, I don't have any time before holidays to fix these issues.  I decided to back out the commit.

I can fix these build problems when I return from holidays in 2 weeks and a bit.   The root pointer concern doesn't seem like a showstopper.  Maybe the PR can be merged and I'll commit a new PR on the root pointer fix when I get back?  Or, just wait...

Reply:
Let's merge.  We can fix it later.